File: DSA/PriyankaSingh_590016827/day82question1.c
#include <stdio.h>
#include <stdlib.h>

// Definition for a binary tree node.
typedef struct TreeNode {
    int data;
    struct TreeNode* left;
    struct TreeNode* right;
} TreeNode;

// Function to create a new tree node
TreeNode* createNode(int data) {
    TreeNode* newNode = (TreeNode*)malloc(sizeof(TreeNode));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Balance BST from Sorted Array
TreeNode* sortedArrayToBST(int arr[], int start, int end) {
    if (start > end) {
        return NULL;
    }
    int mid = start + (end - start) / 2;
    TreeNode* root = createNode(arr[mid]);
    root->left = sortedArrayToBST(arr, start, mid - 1);
    root->right = sortedArrayToBST(arr, mid + 1, end);
    return root;
}

// Helper function to print tree preorder for verification
void preorderTraversal(TreeNode* root) {
    if (root == NULL) {
        return;
    }
    printf("%d ", root->data);
    preorderTraversal(root->left);
    preorderTraversal(root->right);
}

int main() {
    // Example 1
    int arr1[] = {1, 2, 3, 4, 5, 6, 7};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);
    TreeNode* root1 = sortedArrayToBST(arr1, 0, n1 - 1);
    printf("Preorder Traversal: ");
    preorderTraversal(root1);
    printf("\n");

    // Example 2
    int arr2[] = {10, 20, 30, 40};
    int n2 = sizeof(arr2) / sizeof(arr2[0]);
    TreeNode* root2 = sortedArrayToBST(arr2, 0, n2 - 1);
    printf("Preorder Traversal: ");
    preorderTraversal(root2);
    printf("\n");

    return 0;
}

File: DSA/LakshaKundra_590013237/Day_82_Question_1.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def sorted_array_to_bst(arr):
    if not arr:
        return None
    mid = len(arr) // 2
    root = Node(arr[mid])
    root.left = sorted_array_to_bst(arr[:mid])
    root.right = sorted_array_to_bst(arr[mid+1:])
    return root

def preorder(root):
    if root:
        print(root.val, end=" ")
        preorder(root.left)
        preorder(root.right)

arr1 = [1, 2, 3, 4, 5, 6, 7]
root1 = sorted_array_to_bst(arr1)
preorder(root1)
print()

arr2 = [10, 20, 30, 40]
root2 = sorted_array_to_bst(arr2)
preorder(root2)
